cpu/kinetis_common/periph/gpio.c

```diff
@@ -24,384 +24,262 @@
-     * settings later. */
-    PORT_Type *port = _port(dev);
-    uint8_t pin_number = _pin_num(dev);
-    uint32_t irqc = PORT_PCR_IRQC_MASK & port->PCR[pin_number];
-    gpio_int_config_entry_t* entry = NULL;
-    mutex_lock(&int_config_lock);
-    /* Search for the given pin in the port's interrupt configuration */
-    LL_SEARCH_SCALAR(gpio_interrupts[_port_num(dev)], entry, pin, pin_number);
-    if (entry == NULL) {
-        /* Pin has not been configured for interrupts */
-        mutex_unlock(&int_config_lock);
-        return;
-    }
-    entry->irqc = irqc;
-    mutex_unlock(&int_config_lock);
-    port->PCR[pin_number] &= ~(PORT_PCR_IRQC_MASK);
+    int ctx = get_ctx(port_num(pin), pin_num(pin));
+    port(pin)->PCR[pin_num(pin)] |= isr_ctx[ctx].state;
 }
 
-int gpio_read(gpio_t dev)
+void gpio_irq_disable(gpio_t pin)
 {
-    return ((_gpio(dev)->PDIR & GPIO_PDIR_PDI(1 << _pin_num(dev))) ? 1 : 0);
+    int ctx = get_ctx(port_num(pin), pin_num(pin));
+    isr_ctx[ctx].state = port(pin)->PCR[pin_num(pin)] & PORT_PCR_IRQC_MASK;
+    port(pin)->PCR[pin_num(pin)] &= ~(PORT_PCR_IRQC_MASK);
 }
 
-void gpio_set(gpio_t dev)
+int gpio_read(gpio_t pin)
 {
-    _gpio(dev)->PSOR = (1 << _pin_num(dev));
+    if (gpio(pin)->PDDR & (1 << pin_num(pin))) {
+        return (gpio(pin)->PDOR & (1 << pin_num(pin))) ? 1 : 0;
+    }
+    else {
+        return (gpio(pin)->PDIR & (1 << pin_num(pin))) ? 1 : 0;
+    }
 }
 
-void gpio_clear(gpio_t dev)
+void gpio_set(gpio_t pin)
 {
-    _gpio(dev)->PCOR = (1 << _pin_num(dev));
+    gpio(pin)->PSOR = (1 << pin_num(pin));
 }
 
-void gpio_toggle(gpio_t dev)
+void gpio_clear(gpio_t pin)
 {
-    _gpio(dev)->PTOR = (1 << _pin_num(dev));
+    gpio(pin)->PCOR = (1 << pin_num(pin));
 }
 
-void gpio_write(gpio_t dev, int value)
+void gpio_toggle(gpio_t pin)
+{
+    gpio(pin)->PTOR = (1 << pin_num(pin));
+}
+
+void gpio_write(gpio_t pin, int value)
 {
     if (value) {
-        _gpio(dev)->PSOR = (1 << _pin_num(dev));
+        gpio(pin)->PSOR = (1 << pin_num(pin));
     }
     else {
-        _gpio(dev)->PCOR = (1 << _pin_num(dev));
+        gpio(pin)->PCOR = (1 << pin_num(pin));
     }
 }
 
-static inline void irq_handler(uint8_t port_num)
+static inline void irq_handler(PORT_Type *port, int port_num)
 {
-    gpio_int_config_entry_t *entry;
-    PORT_Type *port = _port_ptrs[port_num];
-    uint32_t isf = port->ISFR; /* Interrupt status flags */
-    LL_FOREACH(gpio_interrupts[port_num], entry) {
-        if (isf & (1 << entry->pin)) {
-            if (entry->cb != NULL) {
-                entry->cb(entry->arg);
-            }
+    /* take interrupt flags only from pins which interrupt is enabled */
+    uint32_t status = port->ISFR;
+
+    for (int i = 0; i < 32; i++) {
+        if ((status & (1 << i)) && (port->PCR[i] & PORT_PCR_IRQC_MASK)) {
+            port->ISFR = (1 << i);
+            int ctx = get_ctx(port_num, i);
+            isr_ctx[ctx].cb(isr_ctx[ctx].arg);
```

Comment:
I decided actively against this to keep the ISR routine as short as possible. As the `ctx` is only modified in this driver, the chance of it accidentally being set to `NULL` is zero so I think it is save here to go without the check. The ISR is slow enough already...